Vadinur/profile 2 3

vadimrn opened the pull request.

vadimrn committed:
Реализован функционал экрана "Мои NFT" с UI, логикой и сетевыми запросами

- Добавлена поддержка для отображения списка NFT в UITableView в UserNFTViewController
- Разработана пользовательская ячейка (NFTCell) для отображения деталей NFT
- Реализована кнопка сортировки для управления порядком отображения NFT
- Включена заглушка для сценария отсутствия NFT
- Реализована логика отображения заглушки при отсутствии NFT и функционал сортировки NFT
- Выполнена работа с сетью для загрузки данных "моих NFT" через UserNFTModel и UserNFTViewModel

vadimrn committed:
Реализован основной функционал экрана "Мои NFT" с сетевыми запросами и UI-элементами

- Добавлена поддержка для отображения списка NFT в UITableView в UserNFTViewController
- Разработана пользовательская ячейка (NFTCell) для отображения деталей NFT
- Реализована кнопка сортировки и логика отображения заглушки при отсутствии NFT
- Выполнена работа с сетью для загрузки данных "моих NFT" и информации об авторах через UserNFTModel и UserNFTViewModel
- Добавлены утилитарные расширения для упрощения работы с UI элементами

vadimrn committed:
Добавлены сетевые запросы для получения данных NFT и авторов
- Реализован FetchNFTNetworkRequest для получения данных NFT
- Реализован FetchAuthorNetworkRequest для получения информации об авторах

vadimrn committed:
Реализован универсальный сервис для отображения всплывающих уведомлений

- Добавлены структуры AlertActionModel и AlertModel для конфигурации алертов
- Реализован AlertServiceProtocol с универсальным методом showAlert
- AlertService поддерживает настраиваемые алерты с текстовыми полями и кнопками

vadimrn committed:
Добавлено перечисление SortOption для опций сортировки с локализацией
- Реализовано enum SortOption с вариантами сортировки: price, rating, title
- Добавлены локализованные описания для каждого варианта сортировки

vadimrn committed:
Реализован экран редактирования профиля с функционалом обновления фото и данных

vadimrn committed:
Внесена информация в файл md по декомпозиции о затраченом времени

Kolesnikov-Eugene commented on FakeNFT/Helpers/ViewControllerFactory.swift:
```diff
@@ -3,19 +3,21 @@ import UIKit
     }
 
     func makeEditingViewController(viewModel: ProfileViewModelProtocol) -> EditingViewController {
-        return EditingViewController(viewModel: viewModel)
+        return EditingViewController(viewModel: viewModel as! EditingViewModelProtocol)
```
Лучше разделить ответственности и создавать EditingViewModel отдельно, а не форскастить ProfileViewModel

vadimrn replied:
Исправлено

Kolesnikov-Eugene commented on FakeNFT/Profile/EditScreen/EditingViewModel.swift:
```diff
@@ -0,0 +1,127 @@
+        }
+    }
+    
+    private func fetchUserProfile() {
+        ProgressHUD.show(NSLocalizedString("ProgressHUD.loading", comment: ""))
+        profileService.fetchProfile { [weak self] result in
```
Управлять лоадером лучше следует во вью контроллере

vadimrn replied:
Исправлено

Kolesnikov-Eugene commented on FakeNFT/Profile/ProfileScreen/ProfileViewModel.swift:
```diff
@@ -2,102 +2,108 @@ import Foundation
         $userProfile.observe(handler)
     }
-
-    func fetchUserProfile() {
-        ProgressHUD.show(NSLocalizedString("ProgressHUD.loading", comment: ""))
-
```
Здесь также лучше переместить лоадер во вьюконтроллер

vadimrn committed:
Fix багов

Kolesnikov-Eugene commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+            guard let self = self else { return }
+            
+            switch state {
+            case .loading:
+                print("Загрузка")
```
принты, используемые во время отладки, можно убрать. Лучше добавить необходимую реализацию, либо вообще убрать кейс, если он не используется

DanilaG replied:
Например, тут можно показать лоудер из `ProgressHUD`, чтобы не показывать его во ViewModel)

vadimrn replied:
Исправлено

Kolesnikov-Eugene commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+        let cell: NFTCell = tableView.dequeueReusableCell()
+        cell.selectionStyle = .none
+        
+        guard let nft = viewModel.userNFT?[indexPath.row] else {
+            print("error to get NFT")
+            return cell
```
Еще неиспользуемые принты, можно почистить

vadimrn replied:
Исправлено

vadimrn replied:
исправлено

Kolesnikov-Eugene commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewModel.swift:
```diff
@@ -0,0 +1,122 @@
+        $state.observe(handler)
+    }
+    
+    func fetchNFT(nftList: [String]) {
+        ProgressHUD.show(NSLocalizedString("ProgressHUD.loading", comment: ""))
+        state = .loading
```
Еще один лоадер за пределами контроллера, лучше переместить

vadimrn replied:
Исправлено

Azonaz commented on FakeNFT/Profile/UserNFTScreen/Views/NFTCell.swift:
```diff
@@ -0,0 +1,148 @@
+        
+        if let formattedPrice = NumberFormatter.defaultPriceFormatter.string(from: NSNumber(value: nft.price)) {
+            self.currentPriceLabel.text = "\(formattedPrice) ETH"
+        }
```
Обычно строку формируют вне ячейки и передают готовую строку в ячейку,
форматирование цены тоже должно происходить вне ячейки.
Ячейка - это вьюшка, ее дело просто отобразить те данные которые ей дали =)

vadimrn replied:
Исправлено

vadimrn committed:
Обеспечение потокобезопасности доступа к UserProfile

DanilaG commented on FakeNFT/Foundation/NumberFormatter+Presets.swift:
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+extension NumberFormatter {
+    static var defaultPriceFormatter: NumberFormatter = {
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .decimal
+        formatter.maximumFractionDigits = 2
+        formatter.decimalSeparator = ","
+        return formatter
+    }()
```
🔥 что сделал форматер для суммы + он не пересоздаётся каждый раз

Тут можно только var на let поменять, чтобы случайно не изменить форматер

vadimrn replied:
Исправил

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+    
+    // MARK: - Lifecycle
+    
+    init(nftList: [String], viewModel: UserNFTViewModelProtocol) {
```
`nftList` - нужен только model, и, по логике (так как загрузку осуществляет model), данные элементы должны передоваться в model при ините, а не в view (так как они ни как не связаны с отрисовкой)

Давай поправим этот момент, для этого `nftList` будем передовать на ините в модель, а во view model и в model добавим методы, с помощью которых мы узнаем, что view готова и данные можно грузить. Так как иначе возможен случай, что данные загрузятся до того, как загрузится view, в результате чего произадёт падение.

vadimrn replied:
исправил

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+    
+    // MARK: - UI properties
+    
+    private lazy var alertService: AlertServiceProtocol = {
+        return AlertService(viewController: self)
+    }()
```
Давай инъектить `alertService` через инит, тогда `UserNFTViewController` не будет зависить от конкретной реализации `AlertServiceProtocol `

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell: NFTCell = tableView.dequeueReusableCell()
```
Класное решение)

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell: NFTCell = tableView.dequeueReusableCell()
+        cell.selectionStyle = .none
```
Давай добавим в `NFTCell.setupViews`, чтобы каждый раз не дёргать это поле, а сконфигурировать его единожды

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+            return cell
+        }
+        
+        if let author = viewModel.authors[nft.author] {
+                cell.configure(nft: nft, authorName: author.name)
+            } else {
+                print("error to get author ID")
+                cell.configure(nft: nft, authorName: "Unknown author")
+            }
```
Давай поправим отступы

DanilaG replied:
Чтобы не дублировать вызов функции можно заменить на:
```
        cell.configure(
            nft: nft,
            authorName: viewModel.authors[nft.author]?.name ?? "Unknown author"
        )
```

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+                cell.configure(nft: nft, authorName: author.name)
+            } else {
+                print("error to get author ID")
+                cell.configure(nft: nft, authorName: "Unknown author")
```
"Unknown author" лучше вынести в строки локализации

vadimrn replied:
исправлено

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+    }
+}
+
+// MARK: - UITableViewDelegate
+
+extension UserNFTViewController: UITableViewDelegate {
+    
+}
```
Если делегат не нужен, то давай его удалим)

vadimrn replied:
исправлено

DanilaG commented on FakeNFT/Profile/ProfileScreen/ProfileViewModel.swift:
```diff
@@ -124,9 +129,26 @@ final class ProfileViewModel: ProfileViewModelProtocol {
+    }
+    
+    func fetchUserProfile() {
+        ProgressHUD.show(NSLocalizedString("ProgressHUD.loading", comment: ""))
```
Аналогичный кейс, о котором писал твой коллега)
Давай почистим все ViewModel-и от использования ProgressHUD, так как в этом случае получаем что view model зависит от UI фреймворка

DanilaG commented on FakeNFT/en.lproj/Localizable.strings:
```diff
@@ -19,4 +19,29 @@
 //ProgressHUD
 "ProgressHUD.loading" = "Loading ...";
 
+//NFTCell
+"NFTCell.from" = "from";
+"NFTCell.price" = "Price";
+
+//UserNFTViewController
+"UserNFTViewController.nonft" = "You don't have an NFT yet";
+
+//AlertAction
+"AlertAction.close" = "Close";
+"AlertAction.sort" = "Sorting";
+"AlertAction.ok" = "OK";
+"AlertAction.error" = "Error";
+"AlertAction.incorrURL" = "Incorrect URL entered";
+"AlertAction.cancel" = "Cancel";
+"AlertAction.enterURL" = "Enter URL";
+"AlertAction.imageURL" = "Image URL";
+
+//SortOptions
+"SortOptions.price" = "By price";
+"SortOptions.rate" = "By rating";
+"SortOptions.name" = "By name";
```
Давай добавим в ru локализацию также значения, чтобы при русском языке не показывались ключи

vadimrn replied:
Исправлено

vadimrn committed:
Рефакторинг: Инициализация nftList в модели для оптимизации загрузки данных

vadimrn committed:
Добавлен метод makeEditingViewController в ViewControllerFactory
- Реализован новый метод makeEditingViewController в ViewControllerFactory для создания и инициализации EditingViewController с его ViewModel.
- Обновлен ProfileRouting для использования нового метода makeEditingViewController для маршрутизации к EditingViewController.

vadimrn committed:
Убраны неиспользуемые print

vadimrn committed:
Произведен рефакторинг ячейки NFT: разделение форматирования и отображения данных

vadimrn committed:
Вынесение строки 'Unknown author' в ресурсы локализации

vadimrn committed:
удален не нужный делегат

vadimrn committed:
"Добавление русских переводов для строк локализации"

vadimrn committed:
Управление лоадером перенесено во вью контроллер

vadimrn committed:
Перемещение управления ProgressHUD из ViewModel в ViewController

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,182 @@
+                    self.updateUIBasedOnNFTData()
+                case .error(let error):
+                    ProgressHUD.dismiss()
+                    print("Ошибка: \(error)")
+                    // ToDo: - Показать пользовательский алерт об ошибке
```
Давай добавим отображение ошибки или если ошибки игнорируются, и state с ошибкой ни когда не может произойти, то давай удалим error из enum

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewModel.swift:
```diff
@@ -0,0 +1,126 @@
+    }
+    
+    func fetchNFT(nftList: [String]) {
+     //   ProgressHUD.show(NSLocalizedString("ProgressHUD.loading", comment: ""))
```
Давай удалим не нужные комменты) Это позволит сделать код чище

vadimrn replied:
Исправил

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -164,23 +168,15 @@ extension UserNFTViewController: UITableViewDataSource {
-                cell.configure(nft: nft, authorName: "Unknown author")
-            }
+        if viewModel.authors[nft.author] != nil {
+            cell.configure(nft: nft, authorName: NSLocalizedString("unknownAuthorKey", comment: "Default author name when the actual name is unavailable"))
```
Давай разабъём строчку на несколько, что позволит сделать код более читаемым

vadimrn replied:
Исправлено

vadimrn replied:
Исправлено

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -34,13 +35,13 @@ final class UserNFTViewController: UIViewController {
     
-    private let nftList: [String]
-    private let viewModel: UserNFTViewModelProtocol
+    //  private let nftList: [String]
```
Давай удалим лишний коммент)

vadimrn replied:
Исправлено

DanilaG commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -34,13 +35,13 @@ final class UserNFTViewController: UIViewController {
-    init(nftList: [String], viewModel: UserNFTViewModelProtocol) {
-        self.nftList = nftList
+    init(viewModel: UserNFTViewModelProtocol) {
+   //     self.nftList = nftList
```
Давай удалим лишний коммент)

vadimrn replied:
исправлено

vadimrn committed:
Очистка UserNFTViewModel от ненужных комментариев

vadimrn committed:
Улучшение читаемости кода путем разбиения длинных строк

vadimrn committed:
Удален лишний коментарий

vadimrn committed:
код очищен отлишних коментариев

vadimrn committed:
Удаление error из enum LoadingState
Удалены случаи ошибок из enum LoadingState, так как они не могут произойти. Добавлено отображение ошибок, если они все еще будут игнорироваться.

vadimrn committed:
Отображение метки "Нет NFT" при пустом списке NFT пользователя
- Обновлен метод updateUIBasedOnNFTData для проверки количества userNFT.
- Метка noNFTLabel теперь видима, когда список userNFT пуст или равен nil.
- Обеспечивает дружелюбную обратную связь пользователю, когда нет NFT для отображения.

vadimrn committed:
Исправление отображения имени автора в списке NFT
- Исправлена логика отображения имени автора в ячейках списка NFT.
- Теперь корректно отображается реальное имя автора, если оно доступно.
- В случае отсутствия информации об авторе отображается 'Неизвестный автор'.
